src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -179,13 +181,31 @@ public Task StartAsync(ITransportFactory transportFactory, HttpClient httpClient
             }
         }
 
-        private static async Task<Uri> GetConnectUrl(Uri url, HttpClient httpClient, ILogger logger)
+        private static async Task<NegotiationResponse> Negotiate(Uri url, HttpClient httpClient, ILogger logger)
         {
-            var connectionId = await GetConnectionId(url, httpClient, logger);
-            return Utils.AppendQueryString(url, "id=" + connectionId);
+            var negotiationPayload = await RequestNegotiate(url, httpClient, logger);
+
+            try
+            {
+                using (var reader = new JsonTextReader(new StringReader(negotiationPayload)))
+                {
+                    var negotiationResponse = new JsonSerializer().Deserialize<NegotiationResponse>(reader);
+                    if (negotiationResponse == null)
+                    {
+                        throw new FormatException("Invalid negotiation response received.");
+                    }
+
+                    return negotiationResponse;
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogError("Error parsing negotiation response. {0}", ex);
+                throw new FormatException("Invalid negotiation response received.", ex);
```

Comment:
Kind of - I want to keep the exception thrown by Deserialize. I can however restructure the code to keep this exception and avoid nesting FormatExceptions